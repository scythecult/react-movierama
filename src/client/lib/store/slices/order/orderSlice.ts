import { DEFAULT_CANVAS_HEIGHT, DEFAULT_CANVAS_WIDTH, SeatStateMap } from '../../../constants/common';
import type { Canvas, SeatData, SeatType } from '../../../types/OrderPageData';
import type { WithMiddlewareStateCreator } from '../../appStore';

const toggleSeatStateById = (seats: SeatData[], seatId: number) =>
  seats.map((seat) => {
    if (seat.id === seatId) {
      const isSelected = seat.state === SeatStateMap.SELECTED;
      const nextState = isSelected ? SeatStateMap.FREE : SeatStateMap.SELECTED;

      return {
        ...seat,
        state: nextState,
      };
    }

    return seat;
  });

const calculateCartTotalPrice = (cart: SeatData[]) =>
  cart.reduce((initial, current) => {
    initial += current.price;

    return initial;
  }, 0);

type OrderSliceState = {
  seats: SeatData[] | [];
  canvas: Canvas;
  seatTypes: SeatType[];

  cart: SeatData[];
  cartTotalPrice: number;
};

type OrderSliceActions = {
  setSeats: (seats: SeatData[]) => void;
  setSeatTypes: (seatTypes: SeatType[]) => void;
  setCanvas: (canvas: Canvas) => void;

  addToCart: (seatId: number) => void;
  removeFromCart: (seatId: number) => void;
  clearCart: () => void;
  updateCartTicketType: (seatId: number, ticketTypeId: number) => void;
};

export type OrderSlice = OrderSliceState & OrderSliceActions;

export const createOrderSlice: WithMiddlewareStateCreator<OrderSlice> = (set) => ({
  seats: [],
  canvas: { width: DEFAULT_CANVAS_WIDTH, height: DEFAULT_CANVAS_HEIGHT },
  seatTypes: [],

  cart: [],
  cartTotalPrice: 0,

  setSeats: (seats: SeatData[]) =>
    set(() => ({
      seats,
    })),

  setSeatTypes: (seatTypes: SeatType[]) =>
    set(() => ({
      seatTypes,
    })),

  setCanvas: (canvas: Canvas) => set(() => ({ canvas })),

  addToCart: (seatId) =>
    set(({ seats }) => {
      const nextSeats = toggleSeatStateById(seats, seatId);
      const nextCart = nextSeats.filter((seat) => seat.state === SeatStateMap.SELECTED);
      const nextCartTotalPrice = calculateCartTotalPrice(nextCart);

      return {
        seats: nextSeats,
        cart: nextCart,
        cartTotalPrice: nextCartTotalPrice,
      };
    }),

  removeFromCart: (seatId) =>
    set(({ seats, cart }) => {
      const nextSeats = toggleSeatStateById(seats, seatId);
      const nextCart = cart.filter((cartItem) => cartItem.id !== seatId);
      const nextCartTotalPrice = calculateCartTotalPrice(nextCart);

      return {
        seats: nextSeats,
        cart: nextCart,
        cartTotalPrice: nextCartTotalPrice,
      };
    }),

  clearCart: () =>
    set(({ seats }) => {
      const nextSeats = seats.map((seat) => ({ ...seat, state: SeatStateMap.FREE }));

      return {
        seats: nextSeats,
        cart: [],
        cartTotalPrice: 0,
      };
    }),

  updateCartTicketType: (seatId, ticketTypeId) =>
    set(({ cart }) => {
      const nextCart = cart.map((cartItem) => {
        if (cartItem.id === seatId) {
          const { price, id } = cartItem.seatType.ticketTypes.find((ticketType) => ticketType.id === ticketTypeId)!;

          return {
            ...cartItem,
            ticketTypeId: id,
            price,
          };
        }

        return cartItem;
      });
      const nextCartTotalPrice = calculateCartTotalPrice(nextCart);

      return {
        cart: nextCart,
        cartTotalPrice: nextCartTotalPrice,
      };
    }),
});
