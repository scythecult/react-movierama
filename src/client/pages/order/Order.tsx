// import { MoviePicker } from '../movie-picker/MoviePicker';
import { Layout } from '@/client/lib/components/layout/Layout';
import { Hall } from '@/client/pages/order/hall/Hall';
import { Info } from '@/client/pages/order/info/Info';
import { Legend } from '@/client/pages/order/legend/Legend';
import { Seats } from '@/client/pages/order/seats/Seats';
import styles from './styles.module.css';

export const Order = () => {
  return (
    // TODO Use common "page" component
    <main className={styles.order}>
      <Layout className={styles.orderLayout}>
        <h1>Order</h1>
        <Legend />
        <Hall>
          <Seats />
        </Hall>
        <Info />
      </Layout>
    </main>
  );
};
