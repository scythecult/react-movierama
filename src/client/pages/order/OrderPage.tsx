// import { MoviePicker } from '../movie-picker/MoviePicker';
import { Button } from '@/client/lib/components/button/Button';
import { Layout } from '@/client/lib/components/layout/Layout';
import { Hall } from '@/client/pages/order/hall/Hall';
import { LegendController } from '@/client/pages/order/legend/LegendController';
import { PreCheckController } from '@/client/pages/order/pre-check/PreCheckController';
import { SeatsController } from '@/client/pages/order/seats/SeatsController';
import styles from './styles.module.css';

export type OrderPageProps = {
  totalPrice: number;
}
export const OrderPage = (props: OrderPageProps) => {
  const { totalPrice } = props;
  const isPaymentButtonDisabled = totalPrice === 0;
  const paymentButtonText = isPaymentButtonDisabled ? 'Select seats' : `Proceed to payment: ${totalPrice} RUB`;

  return (
    // TODO Use common "page" component
    <main className={styles.order}>
      <Layout className={styles.orderLayout}>
        <Hall>
          <SeatsController />
        </Hall>
        <LegendController />
        <PreCheckController />
        <div className={styles.orderFooter}>

          <Button disabled={isPaymentButtonDisabled}>{paymentButtonText}</Button>
        </div>

      </Layout>
    </main>
  );
};
