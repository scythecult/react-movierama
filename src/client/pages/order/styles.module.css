.order {
  display: flex;
  flex-direction: column;
  align-items: center;
  gap: 16px;
}

.orderLayout {
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  gap: 18px;
}
