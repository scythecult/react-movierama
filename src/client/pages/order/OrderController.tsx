import { useAppStore } from '@/client/lib/hooks/useAppStore';
import { Order } from './Order';

// Controller get all related data from store and passes it to child components
export const OrderController = () => {
  const seats = useAppStore((state) => state.seats);
  const totalPrice = useAppStore((state) => state.totalPrice);
  const setIsSelected = useAppStore((state) => state.setIsSelected);

  const handleSeatClick = (seatId: string, isSelected: boolean) => {
    setIsSelected(seatId, isSelected);
  };

  return <Order totalPrice={totalPrice} seats={seats} onSeatClick={handleSeatClick} />;
};
