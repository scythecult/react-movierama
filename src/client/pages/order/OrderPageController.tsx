import { useAppStore } from '@/client/lib/hooks/useAppStore';
import { OrderPage } from '@/client/pages/order/OrderPage';

// Controller get all related data from store and passes it to child components
export const OrderPageController = () => {
  const totalPrice = useAppStore((state) => state.totalPrice);

  return <OrderPage totalPrice={totalPrice} />;
};
