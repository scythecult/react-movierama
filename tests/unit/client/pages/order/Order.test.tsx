import { render } from '@testing-library/react';
import { Order } from '@/client/pages/order/Order';

describe('Order', () => {
  test('should correspond default layout', () => {
    const result = render(<Order />);

    expect(result.container).toMatchSnapshot();
  });
});
