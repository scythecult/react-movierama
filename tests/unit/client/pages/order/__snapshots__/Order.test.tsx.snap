// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Order > should correspond default layout 1`] = `
<div>
  <main
    class="_order_e24c8b"
  >
    <div
      class="_layout_762df9 _orderLayout_e24c8b"
    >
      <h1>
        Order
      </h1>
      <ul
        class="_legend_28be33"
      >
        <li
          class="_legendItem_28be33"
        >
          <button
            class="_seat_aa3898 _legendItemSeat_28be33 _disabled_aa3898"
            type="button"
          />
          <small>
            N/A
          </small>
        </li>
        <li
          class="_legendItem_28be33"
        >
          <button
            class="_seat_aa3898 _legendItemSeat_28be33 _selected_aa3898 _disabled_aa3898"
            type="button"
          />
          <small>
            Selected
          </small>
        </li>
        <li
          class="_legendItem_28be33"
        >
          <button
            class="_seat_aa3898 _legendItemSeat_28be33 _occupied_aa3898"
            disabled=""
            type="button"
          />
          <small>
            Occupied
          </small>
        </li>
      </ul>
      <div
        class="_hall_20f54f"
      >
        <div
          class="_hallScreen_20f54f"
        />
        <div
          class="_seats_ba4be0"
        >
          <div
            class="_seatsRow_ba4be0"
          >
            <button
              class="_seat_aa3898"
              type="button"
            />
            <button
              class="_seat_aa3898"
              type="button"
            />
            <button
              class="_seat_aa3898"
              type="button"
            />
            <button
              class="_seat_aa3898"
              type="button"
            />
            <button
              class="_seat_aa3898"
              type="button"
            />
            <button
              class="_seat_aa3898"
              type="button"
            />
            <button
              class="_seat_aa3898"
              type="button"
            />
            <button
              class="_seat_aa3898"
              type="button"
            />
            <button
              class="_seat_aa3898"
              type="button"
            />
            <button
              class="_seat_aa3898"
              type="button"
            />
            <button
              class="_seat_aa3898 _occupied_aa3898"
              disabled=""
              type="button"
            />
            <button
              class="_seat_aa3898 _occupied_aa3898"
              disabled=""
              type="button"
            />
            <button
              class="_seat_aa3898"
              type="button"
            />
            <button
              class="_seat_aa3898"
              type="button"
            />
          </div>
          <div
            class="_seatsRow_ba4be0"
          >
            <button
              class="_seat_aa3898"
              type="button"
            />
            <button
              class="_seat_aa3898"
              type="button"
            />
            <button
              class="_seat_aa3898"
              type="button"
            />
            <button
              class="_seat_aa3898"
              type="button"
            />
            <button
              class="_seat_aa3898"
              type="button"
            />
            <button
              class="_seat_aa3898"
              type="button"
            />
            <button
              class="_seat_aa3898"
              type="button"
            />
            <button
              class="_seat_aa3898"
              type="button"
            />
            <button
              class="_seat_aa3898"
              type="button"
            />
            <button
              class="_seat_aa3898"
              type="button"
            />
            <button
              class="_seat_aa3898"
              type="button"
            />
            <button
              class="_seat_aa3898"
              type="button"
            />
            <button
              class="_seat_aa3898"
              type="button"
            />
            <button
              class="_seat_aa3898"
              type="button"
            />
          </div>
          <div
            class="_seatsRow_ba4be0"
          >
            <button
              class="_seat_aa3898"
              type="button"
            />
            <button
              class="_seat_aa3898"
              type="button"
            />
            <button
              class="_seat_aa3898"
              type="button"
            />
            <button
              class="_seat_aa3898"
              type="button"
            />
            <button
              class="_seat_aa3898"
              type="button"
            />
            <button
              class="_seat_aa3898 _occupied_aa3898"
              disabled=""
              type="button"
            />
            <button
              class="_seat_aa3898 _occupied_aa3898"
              disabled=""
              type="button"
            />
            <button
              class="_seat_aa3898"
              type="button"
            />
            <button
              class="_seat_aa3898"
              type="button"
            />
            <button
              class="_seat_aa3898"
              type="button"
            />
            <button
              class="_seat_aa3898"
              type="button"
            />
            <button
              class="_seat_aa3898"
              type="button"
            />
            <button
              class="_seat_aa3898"
              type="button"
            />
            <button
              class="_seat_aa3898"
              type="button"
            />
          </div>
          <div
            class="_seatsRow_ba4be0"
          >
            <button
              class="_seat_aa3898 _occupied_aa3898"
              disabled=""
              type="button"
            />
            <button
              class="_seat_aa3898"
              type="button"
            />
            <button
              class="_seat_aa3898"
              type="button"
            />
            <button
              class="_seat_aa3898"
              type="button"
            />
            <button
              class="_seat_aa3898"
              type="button"
            />
            <button
              class="_seat_aa3898"
              type="button"
            />
            <button
              class="_seat_aa3898"
              type="button"
            />
            <button
              class="_seat_aa3898"
              type="button"
            />
            <button
              class="_seat_aa3898"
              type="button"
            />
            <button
              class="_seat_aa3898"
              type="button"
            />
            <button
              class="_seat_aa3898"
              type="button"
            />
            <button
              class="_seat_aa3898"
              type="button"
            />
            <button
              class="_seat_aa3898"
              type="button"
            />
            <button
              class="_seat_aa3898"
              type="button"
            />
          </div>
          <div
            class="_seatsRow_ba4be0"
          >
            <button
              class="_seat_aa3898"
              type="button"
            />
            <button
              class="_seat_aa3898"
              type="button"
            />
            <button
              class="_seat_aa3898"
              type="button"
            />
            <button
              class="_seat_aa3898"
              type="button"
            />
            <button
              class="_seat_aa3898 _occupied_aa3898"
              disabled=""
              type="button"
            />
            <button
              class="_seat_aa3898"
              type="button"
            />
            <button
              class="_seat_aa3898"
              type="button"
            />
            <button
              class="_seat_aa3898"
              type="button"
            />
            <button
              class="_seat_aa3898"
              type="button"
            />
            <button
              class="_seat_aa3898"
              type="button"
            />
            <button
              class="_seat_aa3898"
              type="button"
            />
            <button
              class="_seat_aa3898"
              type="button"
            />
            <button
              class="_seat_aa3898"
              type="button"
            />
            <button
              class="_seat_aa3898"
              type="button"
            />
          </div>
        </div>
      </div>
      <p
        class="_info_1a8a63"
      >
        You have selected 
        <span
          class="_infoText_1a8a63"
        >
          0
        </span>
         seats for a price of
        <span
          class="_infoText_1a8a63"
        >
           $
          0
        </span>
      </p>
    </div>
  </main>
</div>
`;
